### PR TITLE
2.x: implement concat(Map)Eager(DelayError) and expose concat(Map)DelayError overload

### PR DESCRIPTION
**Pull request by akarnokd**
Started a syncing process between 1.x and 2.x operators (based on [this diff](https://gist.github.com/akarnokd/4a391527e099412cad87d3dffbfef762)). The first set is the addition of the eager version of `concat` and `concatMap`. In addition, all of them get a `XDelayError` overloads as well.

### CODE REVIEW

**Comment by akarnokd** [comment]
Also I've expanded Flowable with a `test()` method. Should give a nice answer to the question "How do you test a Flowable?".

File: src/main/java/io/reactivex/internal/util/QueueDrainHelper.java [hunk 1]
```diff
@@ -359,4 +360,29 @@
...
+    /**
+     * Creates a queue: spsc-array if capacityHint is positive and 
+     * spsc-linked-array if capacityHint is negative; in both cases, the
+     * capacity is the absolute value of prefetch.
+     * @param <T> the value type of the queue
+     * @param capacityHint the capacity hint
+     * @return the queue instance
+     */
+    public static <T> Queue<T> createQueue(int capacityHint) {
+        if (capacityHint < 0) {
```
Comment (vanniktech) [inline src/main/java/io/reactivex/internal/util/QueueDrainHelper.java:373]: how about `return new SpscLinkedArrayQueue<T>(Math.abs(capacityHint));`

Comment (akarnokd) [inline src/main/java/io/reactivex/internal/util/QueueDrainHelper.java:373]: From the JDK:

``` java
public static long abs(long a) {
        return (a < 0) ? -a : a;
}
```

Pointless, plus this is an internal utility method.

Comment (vanniktech) [inline src/main/java/io/reactivex/internal/util/QueueDrainHelper.java:373]: In my opinion it makes reading the code a bit easier and 3 lines will be saved too but yeah it's just a nit.

File: src/test/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEagerTest.java [hunk 2]
```diff
@@ -230,4 +230,20 @@ public void mainErrorsImmediate() {
         
         ts.assertFailureAndMessage(TestException.class, "Forced failure", 2);
     }
+    
+    @Test
+    public void longEager() {
+
```
Comment (artem-zinnatullin) [inline src/test/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEagerTest.java:236]: nit: el

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEager.java [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    protected void subscribeActual(Subscriber<? super R> s) {
+        source.subscribe(new ConcatMapEagerDelayErrorSubscriber<T, R>(
+                s, mapper, maxConcurrency, prefetch, errorMode));
+    }
+
+    static final class ConcatMapEagerDelayErrorSubscriber<T, R>
+    extends AtomicInteger
+    implements Subscriber<T>, Subscription, InnerQueuedSubscriberSupport<R> {
+        /** */
+        private static final long serialVersionUID = -4255299542215038287L;
```
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEager.java:64]: nit: turn off IDE warning and remove `serialVersionUID`?

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEager.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                s.cancel();
+                onError(ex);
+                return;
+            }
+            
+            InnerQueuedSubscriber<R> inner = new InnerQueuedSubscriber<R>(this, prefetch);
+            
+            if (!cancelled) {
+                subscribers.offer(inner);
+                if (!cancelled) {
```
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEager.java]: nit: can we invert that and make `drainAndCancel` first so it'll be easier to read and harder to make some mistake during refactoring?

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEager.java [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+        public void request(long n) {
+            if (SubscriptionHelper.validate(n)) {
+                BackpressureHelper.add(requested, n);
+                drain();
+            }
+        }
+        
+        @Override
+        public void innerNext(InnerQueuedSubscriber<R> inner, R value) {
+            if (!inner.queue().offer(value)) {
```
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEager.java]: nit: same here about reversing `if-else`. I think PMD complains about it btw

**Comment by akarnokd** [comment]
Updated the `if`s

File: src/main/java/io/reactivex/internal/subscribers/flowable/InnerQueuedSubscriber.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+import io.reactivex.internal.subscriptions.SubscriptionHelper;
+import io.reactivex.internal.util.QueueDrainHelper;
+
+/**
+ * Subscriber that can fuse with the upstream and calls a support interface
+ * whenever an event is available.
+ *
+ * @param <T> the value type
+ */
+public final class InnerQueuedSubscriber<T>
```
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/internal/subscribers/flowable/InnerQueuedSubscriber.java:31]: Maybe add separate unit tests for it and check that behavior of public API is correct and it correctly reacts to different fusion modes? Up to you of course

Comment (akarnokd) [inline src/main/java/io/reactivex/internal/subscribers/flowable/InnerQueuedSubscriber.java:31]: Will be part of the test-sync and coverage improvement before RC1.

File: src/main/java/io/reactivex/subscribers/TestSubscriber.java [hunk 7]
```diff
@@ -831,14 +831,58 @@ private String fusionModeToString(int mode) {
...
+    }
+
+    /**
+     * Assert that the upstream signalled the specified values in order
+     * and then failed with a specific class or subclass of Throwable.
+     * @param error the expected exception (parent) class
+     * @param values the expected values, asserted in order
+     * @return this
+     */
+    public final TestSubscriber<T> assertFailure(Class<Throwable> error, T... values) {
```
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/subscribers/TestSubscriber.java:862]: nice

File: src/main/java/io/reactivex/internal/subscribers/flowable/InnerQueuedSubscriber.java (hunk 6 above)
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/internal/subscribers/flowable/InnerQueuedSubscriber.java:31]: ok

File: src/test/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEagerTest.java [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    @Test
+    public void normal() {
+        Flowable.range(1, 5)
+        .concatMapEager(new Function<Integer, Publisher<Integer>>() {
+            @Override
+            public Publisher<Integer> apply(Integer t) {
+                return Flowable.range(t, 2);
+            }
+        })
+        .test()
```
Comment (artem-zinnatullin) [inline src/test/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEagerTest.java:37]: oh man, this is awesome 👍

**Comment by artem-zinnatullin** [comment]
I'm thinking if there is a way to test such [`if (!cancelled)`](https://github.com/ReactiveX/RxJava/pull/4190/commits/2bf22b287c78c49709db098aea2dcd4ca858a5a1) errors, java debuggers allows suspending a Thread on a particular breakpoint and allow other threads to run at the same time, maybe such mechanism will allow us to simulate and test concurrent access in particular parts of code (drain loops and so on). 

Something like:

``` java
@Test
public void concurrentCancel() {
  TestThread thread1 = testThread({
     disposable = Flowable.just(1).concatMap(v -> Flowable.just(2)).subscribe();
  });

  TestThread thread2 = testThread({
    disposable.dispose();
  });

  TestThreads.builder()
    .start(thread1) // One or multiple.
    .thenSuspend(thread1, FlowableConcatMapEager.class, 128) // Thread; Source class; Line.
    .thenStart(thread2)
    .thenResume(thread1)
    .run(); // Run and wait all test threads to complete.

  // Assert behavior.
}
```

Just thoughts

**Comment by artem-zinnatullin** [comment]
:+1: for PR

**Comment by akarnokd** [comment]
Thanks @artem-zinnatullin !

**Pull request merged**